File: python/future_city.py
INF=987654321
n,m=map(int,input().split())
adj= [[INF]*(n+1) for _ in range(n+1)]
for i in range(1,n+1):
    adj[i][i]=0
for _ in range(m):
    a,b=map(int,input().split())
    adj[a][b]=1
    adj[b][a]=1

for k in range(1,n+1):
    for i in range(1,n+1):
        for j in range(1,n+1):
            adj[i][j]=min(adj[i][j],adj[i][k]+adj[k][j])

x,k=map(int,input().split())
ret= adj[1][k]+adj[k][x]
if ret>=INF:
    print(-1)
else:
    print(ret)

File: c++/future_city.cpp
#include <iostream>
#include <vector>
using namespace std;

const int INF = 987654321;
int adj[101][101];
int n, m;
void solve() {
	cin >> n >> m;
	for(int i=1;i<=n;i++)
		for (int j = 1; j <= n; j++) {
			if (i == j)continue;
			adj[i][j] = INF;
		}
	for (int i = 0; i < m; i++) {
		int a, b;
		cin >> a >> b;
		adj[a][b] = 1;
		adj[b][a] = 1;
	}
	for (int k = 1; k <= n; k++)
		for (int i = 1; i <= n; i++)
			for (int j = 1; j <= n; j++)
				adj[i][j] = min(adj[i][j], adj[i][k] + adj[k][j]);
	int x, k;
	cin >> x >> k;
	int ret = adj[1][k] + adj[k][x];
	if (ret >= INF)
		cout << -1 << "\n";
	else
		cout << ret << "\n";
}

int main() {
	solve();
}